scrapy/utils/item.py

```diff
@@ -0,0 +1,69 @@
+from scrapy.item import BaseItem
+
+
+def is_item_like(obj):
+    """
+    Returns True if *obj* is considered a Scrapy *item*, False otherwise.
+
+    An object is considered an *item* if it is:
+    - a scrapy.item.BaseItem or dict instance (or any subclass)
+    - a dataclass object
+    """
+    return isinstance(obj, (BaseItem, dict)) or _is_dataclass_instance(obj)
+
+
+def _is_dataclass_instance(obj):
+    """
+    Returns True if *obj* is a dataclass object, False otherwise.
+    Taken from https://docs.python.org/3/library/dataclasses.html#dataclasses.is_dataclass.
+    """
+    try:
+        from dataclasses import is_dataclass
+    except ImportError:
+        return False
+    else:
+        return is_dataclass(obj) and not isinstance(obj, type)
+
+
+class ItemAdapter:
```

Comment:
A good idea to have such class!

Do you think we can make it nicer - e.g. use a more dict-like API for it, or even MutableMapping? E.g. why .get_item_field, and not just .get? Why .set_item_field, and not `["key"] = value`? This could allow to reduce usages of as_dict

Reply:
Thanks for the update @elacuesta! I still wonder if we should implement MutableMapping interface. It requires to implement 5 methods: `__getitem__`, `__setitem__`, `__delitem__`, `__iter__`, `__len__`; after that you get `__contains__`, `keys`, `items`, `values`, `get`, `__eq__`, `__ne__`, `pop`, `popitem`, `clear`, `update`, and `setdefault` automatically.

Reply:
And one more thing: do you see any issues with allowing to yield ItemAdaptor instances as items? Do you see any use case for that?

Reply:
Implemented the `MutableMapping` interface, that simplified a few things (no need for `asdict` and `__contains__` anymore, for instance) 🥳 
Just as a note, I see no other way to implement `__delitem__` but to use [`delattr`](https://docs.python.org/3/library/functions.html#delattr), which seems to leave `dataclass` objects in an unstable state, but this also happens if we define our own `__init__` method:

```python
>>> from dataclasses import dataclass, fields, asdict
>>> @dataclass(init=False)
... class TestItem:
...     name: str
...     value: int
...     def __init__(self, name):
...         self.name = name
...
>>> i = TestItem(name="asdf")
>>>
>>> [f.name for f in fields(i)]  # all good here
['name', 'value']
>>> i.name
'asdf'
>>> i.value  # expected to raise AttributeError
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'TestItem' object has no attribute 'value'
>>>
>>> i  # __repr__ breaks
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/dataclasses.py", line 357, in wrapper
    result = user_function(self)
  File "<string>", line 2, in __repr__
AttributeError: 'TestItem' object has no attribute 'value'
>>>
>>> asdict(i)  # this breaks too
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/dataclasses.py", line 1044, in asdict
    return _asdict_inner(obj, dict_factory)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/dataclasses.py", line 1051, in _asdict_inner
    value = _asdict_inner(getattr(obj, f.name), dict_factory)
AttributeError: 'TestItem' object has no attribute 'value'
>>>
```

Reply:
Regarding allowing `ItemAdapter` objects as items, I _think_ there should be no problems, it should be only a matter of allowing it in `is_item_like`.
If we want to avoid the extra wrapping layer that it would create, we could add some `classmethod` to the class in order to return the object itself if it's already an adapter (I did some experiments with `__new__` but that didn't work). Or perhaps we don't really care about that. What do you think?

Reply:
> And one more thing: do you see any issues with allowing to yield ItemAdaptor instances as items? Do you see any use case for that?

I think we should consider preventing this, either by automatically _unwrapping_ items before moving them forward, or by raising an exception.

At the moment, if a user yields an item of a certain type from a spider callback, that user can trust that the item will be the same object the user receives in any custom middleware or pipelines. We lose this if third-party middlewares are allowed to yield these adapter objects and not just use them.

I am not completely against it, but if we want to go this way, it may be better to just convert any item into an Item instance.

Reply:
As for delitem, I think that's fine to raise NonImplementedError - but it seems you already implemented it :)

Reply:
Alright, now I think we should not allow yielding adapters. It reminds me of the question of allowing to yield item loaders, the implementation should not be super hard but it doesn't add much IMHO and it's probably not necessary.
Regarding `__delitem__`, I didn't think about `NonImplementedError` :thinking: 
Sounds as reasonable as using `delattr` to me, I don't have a strong opinion on the subject.